File: linked_list/list_node.cc
#include "list_node.h"

// 注意参数的设置
void AddToTail(ListNode*& head, int val) {
  ListNode* new_node = new ListNode(val);
  if (head == nullptr) {
    head = new_node;
  } else {
    ListNode* curr = head;
    while (curr->next) {
      curr = curr->next;
    }
    curr->next = new_node;
  }
}

void RemoveNode(ListNode*& head, int val) {
  if (head == nullptr) {
    return;
  }

  ListNode* to_be_deleted = nullptr;
  if (head->val == val) {
    to_be_deleted = head;
    head = head->next;
  } else {
    ListNode* curr = head;
    while (curr->next && curr->next->val != val) {
      curr = curr->next;
    }
    if (curr->next && curr->next->val == val) {
      to_be_deleted = curr->next;
      curr->next = curr->next->next;
    }
  }

  if (to_be_deleted) {
    delete to_be_deleted;
    to_be_deleted = nullptr;
  }
}

File: linked_list/list_node.py
from typing import Optional, List


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def create_list_node(nums: List) -> Optional[ListNode]:
    dummy = ListNode(0)
    curr = dummy

    for num in nums:
        curr.next = ListNode(num)
        curr = curr.next

    return dummy.next


def create_cycle_list(nums: List, pos: int) -> Optional[ListNode]:
    dummy = ListNode(0)
    curr = dummy
    cycle_node = None
    for i in range(len(nums)):
        curr.next = ListNode(nums[i])
        curr = curr.next
        if i == pos:
            cycle_node = curr
        if i == len(nums) - 1:
            curr.next = cycle_node

    return dummy.next


# 2026/05/13: 注意 .join 的用法.
def print_list_node(head: ListNode) -> None:
    vals = []
    curr = head
    while curr:
        vals.append(str(curr.val))
        curr = curr.next
    vals.append("N")
    print(" -> ".join(vals))
